report/common.py

```diff
@@ -0,0 +1,455 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Common libraries for report generation."""
+
+import dataclasses
+import io
+import json
+import logging
+import os
+import re
+from typing import List, Optional
+
+import yaml
+from google.cloud import storage
+
+import run_one_experiment
+from experiment import evaluator
+from experiment.workdir import WorkDirs
+
+MAX_RUN_LOGS_LEN = 16 * 1024
+
+TARGET_EXTS = ('.c', '.cc', '.cpp', '.cxx', '.c++', '.java', '.py')
+
+
+@dataclasses.dataclass
+class Benchmark:
+  """The class of a benchmark function and its experiment results."""
+  id: str
+  status: str
+  result: run_one_experiment.AggregatedResult
+  signature: str = ''
+  project: str = ''
+  function: str = ''
+
+
+@dataclasses.dataclass
+class Sample:
+  """Result of a fuzz target sample of a benchmark."""
+  id: str
+  status: str
+  result: evaluator.Result
+
+  @property
+  def stacktrace(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return f'{reproducer_link}/stacktrace'
+
+  @property
+  def target_binary(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return f'{reproducer_link}/target_binary'
+
+  @property
+  def reproducer(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return f'{reproducer_link}/artifacts'
+
+  @property
+  def run_log(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return reproducer_link.removesuffix('reproducer') + 'run.log'
+
+
+@dataclasses.dataclass
+class Target:
+  code: str
+  fixer_prompt: Optional[str] = None
+
+
+class FileSystem:
+  """
+  FileSystem provides a wrapper over standard library and GCS client and
+  automatically chooses which to use based on the provided path.
+  """
+
+  _gcs_client = None
+
+  @classmethod
+  def _get_gcs_client(cls):
+    """
+    Returns a cached storage client (a new one is created on first call.
+
+    A new client does authentication on first call, so caching the client will
+    same multiple authentication round trips to GCP.
+    """
```

Comment:
Thanks for always making the comments so informative!

Reply:
No worries :)